components/builder-depot/src/server.rs

```diff
@@ -1779,12 +1691,50 @@ fn show_package(req: &mut Request) -> IronResult<Response> {
                 session_id,
                 &ident.get_origin(),
             ));
-            request.set_ident(ident);
+            request.set_ident(ident.clone());
+            let u = helpers::extract_query_value("upstream", req);
+
+            // If the "upstream" query parameter is not passed, then we can assume it's a normal
+            // API request and we should check for an update in the background.
+            if u.is_none() {
+                refresh_package_from_upstream_depot(req, ident.clone(), Some(channel.clone()));
+            }
 
             match route_message::<OriginChannelPackageLatestGet, OriginPackageIdent>(req, &request)
             {
                 Ok(id) => ident = id.into(),
-                Err(err) => return Ok(render_net_error(&err)),
+                Err(err) => {
+                    let code = err.get_code();
+
+                    // If the package is not found and it's the hab CLI making the request and
+                    // upstream=t is passed in the query string, we can assume someone is trying to
+                    // hab pkg install this particular package, and we want to synchronously check
+                    // the upstream depot before returning a 404 (assuming an upstream depot is
+                    // configured)
+                    if config.upstream_depot.is_some() && code == ErrCode::ENTITY_NOT_FOUND
+                        && helpers::is_request_from_hab(req)
+                        && u.unwrap_or("f".to_string()) == "t"
+                    {
+                        let depot_cli = depot_client_from_config(&config).unwrap();
```

Comment:
It would be nice if we could encapsulate the use of the depot client to only the refresh code (which could become it's own module/struct - more encapsulated and portable).